### DISCUSSION TITLE
Implement a new API method to limit the number of returned list object

### DISCUSSION THREAD

**Discussion by mkanet** [nicolargo/glances#2491] [Q&A] [answered]
Is there a way to limit the number of items monitored by processlist plugin?  I only need the API to return the top 5 processes.  Doing this should significantly reduce glances CPU consumption when not all processes are relevant.

If this is not possible, what are the chances we could have this feature added to Glances?

**Comment by nicolargo** [comment, marked as answer]
Hi @mkanet 

interesting question you have :)

In fact Glances already optimize the processlist plugin by doing the following:
- get all the processes stats with the sorted key (for example CPU consumption)
-  sort them
-  take the first N processes (N depends of the screen height)
- get extended stats only for those N processes

If i understand, you want to make the N "number" configurable ?

**Reply by mkanet** [reply]
@nicolargo Yes, it would be awesome to.make N configurable... especially to limit size of API respond when processlist is included. 

_NOTE_:  If this is added as a feature and enabled, is it possible you to also limit any API RESTful queries that include processlist; even whe doing a RESTful `/all` query?

**Comment by mkanet** [comment]
> 

Hi @nicolargo I just modified `glances_processlist.py` in my currently installed Glances module; hardcoding `N`; to the value of `5`:

`glances_processlist.py`:
Modifying line 158 from:
```
self.stats = stats
```
to:
```
self.stats = stats[:5]
```
**Now, the API returns only the top 5 items in processlist; reducing the API payload significantly**.  However, I want this to ONLY affect the API, not the Glances webserver GUI!  I want the webserver GUI to still only display all the processes based on the browser's screen height.  I am hoping I can convince you to add a feature to make "N" configurable **for ONLY the API, not GUI.**

Thank you for all your hard work!  Looking forward to the PR.

**Comment by nicolargo** [comment]
Implementation done in the develop branch  https://github.com/nicolargo/glances/issues/2528
Will be release in Glances 4.0.0.
Documentation of the new API entry point here: https://github.com/nicolargo/glances/blob/develop/docs/api.rst#get-top-n-items-of-a-specific-plugin

Thanks for the issue request @mkanet !

**Comment by mkanet** [comment]
> Implementation done in the develop branch #2528 Will be release in Glances 4.0.0. Documentation of the new API entry point here: https://github.com/nicolargo/glances/blob/develop/docs/api.rst#get-top-n-items-of-a-specific-plugin
> 
> Thanks for the issue request @mkanet !

@nicolargo  Thank you very much!  I am curious how would I make a curl API query so that it returns data from **ALL** available plugins _AND_ also limit processlist?

Currently, I use the following command line with the small modification I made to: `glances_processlist.py` (hardcoded for top 5):
```
curl http://192.168.1.2:61208/api/3/all
```
I know your documentation says: _Return a very big dictionary (avoid using this request, performances will be poor)_...  However, that is mostly because of `processlist` can be VERY big!  `curl http://192.168.1.2:61208/api/3/all` is actually NOT poor performance when processlist is limited.  It is actually very good performance!
